### ISSUE TITLE
Bug: [no-unnecessary-type-constraint] False positive for generics on arrow functions in .mts files

### ISSUE THREAD

**Issue by spalger** [typescript-eslint/typescript-eslint#5449]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Issue Description

When using the new ([as of TS 4.7](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-7.html#new-file-extensions)) `.mts` file extension the `const foo = <T>() => ...` syntax for generic arguments is reserved similar to JSX files. The suggestion from tsc recommends defining an explicit constraint:

<img width="1633" alt="image" src="https://user-images.githubusercontent.com/1329312/183544789-01f13bf7-6b1b-4d1c-bf9c-140aff9613e5.png">

Since my function doesn't need or nor do I want a type constraint, but I need one for the syntax, I reached for the noop constraint of `extends any`, but then I was violating the `no-unnecessary-type-constraint` rule.

<img width="1785" alt="image" src="https://user-images.githubusercontent.com/1329312/183545544-05e4430b-c230-4196-a326-d691d0dc5204.png">

In this case, with the `.mts` extension and it's limited syntax I think it's unfair to say it "doesn't do anything and is unnecessary".

### Reproduction Repository Link

https://github.com/spalger/reproduce-issues/tree/ts-eslint-no-unnecessary-type-constraint

### Repro Steps

1) clone the repo with `--branch`

    `git clone git@github.com:spalger/reproduce-issues.git --branch ts-eslint-no-unnecessary-type-constraint`

2) `npm i`
3) `npm run lint`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.33.0` |
| `@typescript-eslint/parser`            | `5.33.0` |
| `TypeScript`                           | `4.7.4` |
| `ESLint`                               | `8.21.0` |
| `node`                                 | `18.7.0` |

**Comment by JoshuaKGoldberg** [comment]
Somehow, I had a feeling this line of code would come back to haunt us 😄:

https://github.com/typescript-eslint/typescript-eslint/blob/48c4ed6c2cddad6d06be9713707357d9983a5b02/packages/eslint-plugin/src/rules/no-unnecessary-type-constraint.ts#L62

Agreed, this is a bug. The rule should understand `.cts` and `.mts` files to need the constraint too.

**Comment by bradzacher** [comment]
We should probably somehow export the utils here to reuse them so we can share logic across packages?

https://github.com/typescript-eslint/typescript-eslint/blob/399b3b1263665288316be6faff759cf514e24b1b/packages/typescript-estree/src/create-program/getScriptKind.ts#L4-L48

**Comment by viell-dev** [comment]
TS7060:
> This syntax is reserved in files with the .mts or .cts extension. Add a trailing comma or explicit constraint.

A trailing comma should also be fine. Like so:
```ts
const func = <T,>(name: T) => {
  console.log(name);
}
```
Though if you're using a formatter, it might end up removing the trailing comma.

**Comment by pkyeck** [comment]
hey, I'm wondering what the status on this is? same happens with `.cts` files.

```typescript
(type parameter) T in <T>(baseConfig: T, process: NodeJS.Process): T
This syntax is reserved in files with the .mts or .cts extension. Add a trailing comma or explicit constraint.ts(7060)
```

![error-cts](https://user-images.githubusercontent.com/870980/220101780-755a75ec-2765-4c6d-8d4d-941196f196ae.png)

**Comment by bradzacher** [comment]
With any issue opened in this project — it either has visible progress in the form of an attached PR, or it has no progress.

We are a community-run project. The volunteer maintainers spend most of their time triaging issues and reviewing PRs. This means that most issues will not progress unless a member of the community steps up and champions it.

If this issue is important to you — consider being that champion.

If not — please just subscribe to the issue and wait patiently.
Commenting asking for status updates does not bump issue priority in any way and just serves to spam everyone subscribed to the issue.

**Comment by WoodyWoodsta** [comment]
@bradzacher This is all very well, but the reality is that these packages require significant amounts of context before approaching being the "champion".

This is an issue that's cropped up for me as well, and I've spent a fair amount of my time trying to understand why this issue even exists in the first place. I'm still not anywhere close to understanding what needs to change in order to fix.

Any help would be appreciated? I'm more than happy to open a pr.

If an issue is labelled `good-first-issue`, I'd expect either a rough sketch of the cause of the issue, or of the fix required, or it to be a relatively trivial issue.

**Comment by bradzacher** [comment]
@WoodyWoodsta [Josh was pretty succinct in his analysis of the problem in the very first comment on this thread](https://github.com/typescript-eslint/typescript-eslint/issues/5449#issuecomment-1209495967)

`.mts` files are seen as JSX files - but the rule was written when `.tsx` files were the only JSX files - thus it gates the rule's JSX logic behind the file's extension - which is obviously now incorrect.

[My second comment in this thread](https://github.com/typescript-eslint/typescript-eslint/issues/5449#issuecomment-1210073224) suggested a utility we could expose from another one of our packages to use instead of the current and incorrect check.

So the PR to fix this would be
1) expose the linked utility from `typesript-estree` - probably through the `type-utils` package.
2) replace the incorrect check with this utility that will behave correctly.

**Comment by WoodyWoodsta** [comment]
@bradzacher Appreciate the response - Josh was very succinct about the symptom, but to create the link between that and the actual problem wasn't as easy. Instead the member analysis was "this came back to bite us" and "We should probably somehow export the utils here..." which doesn't help anyone understand what you've crucially highlighted which is that somehow `mts` files are regarded as JSX files. I still don't understand how that's possible when the highlighted line is `.startsWith('tsx')` but I'm willing to replace that with the utility. Thank you for the input.

**Comment by bradzacher** [comment]
To clarify - in JSX files the single generic argument arrow function syntax (`<T>() => {}`) is considered "syntactically ambiguous" by TS - it needs something to disambiguate the `<T>` as a generic parameter definition as opposed to an opening JSX tag. That disambiguation can either be `<T,>` or `<T extends something>`.

Hence the correct fix in a JSX file is from `<T extends any>` to `<T,>`.

When TS introduced `.mts` files the decision was made that those files would be treated as JSX files by TS (off topic but I'd say that decision was made to avoid needing to introduce `.mts`, `.mtsx`, `.cts`, and `.ctsx` - they dropped the `x` and halved the number of new extensions added).

So hence there's a problem in the rule. The rule was built in a time when the only JSX file was `.tsx` - hence it gates its JSX logic behind the hardcoded extension `.tsx`. This hardcoded string is (as mentioned) no longer correct.

So the options to solve this are two fold:
1) add more hardcoded conditions for the new extensions and hope no new extensions are added in the future (bad).
2) remove the hardcoded check and use the same utility the parser uses to check if the file is a JSX file (good).

**Comment by WoodyWoodsta** [comment]
The background of JSX tags and generic argument syntax isn't new to me, that makes complete sense.

To clarify, is https://github.com/typescript-eslint/typescript-eslint/blob/8ee0cbb0818392cc818b1c254dcec6e879f79a35/packages/utils/src/eslint-utils/rule-tester/RuleTester.ts#L127 the actual source of the issue?

But I think I have enough understanding for a fix. Thank you!

**Comment by WoodyWoodsta** [comment]
Sorry @bradzacher, my snippet I linked was from the jest rule tester, not the actual plugin code. Can't digest an entire monorepo in one day.

Let's continue the discussion in #6795

**Comment by JoseLion** [comment]
I'm seeing this issue on `.tsx` files but with some differences:
- Autofix does not seem to change `const bar = <T extends any>(t: T): T => t;` into `const bar = <T,>(t: T): T => t;`
- After changing to the correct fix, both the `comma-dangle` and `comma-spacing` rules report a problem.

Are these related, or should I open a separate issue?
![image](https://github.com/typescript-eslint/typescript-eslint/assets/3087228/a0647e7c-498d-4277-bf5f-c51f295ade56)
![image](https://github.com/typescript-eslint/typescript-eslint/assets/3087228/e0732c18-feb5-4fd9-b288-26b7efb1f727)

#### Versions:
```json
"@typescript-eslint/eslint-plugin": "^5.61.0",
"@typescript-eslint/parser": "^5.61.0",
"eslint": "^8.44.0",
```

#### Rules Config:
```json
"@typescript-eslint/comma-dangle": ["error", "always-multiline"],
"@typescript-eslint/no-unnecessary-type-constraint": "error",
"comma-spacing": "error",
```

**Comment by bradzacher** [comment]
> After changing to the correct fix, both the comma-dangle and comma-spacing rules report a problem.

Those are separate issues.
